src/client/components/searchBar.ml

```diff
@@ -2,25 +2,73 @@ open Nes
 open Js_of_ocaml
 open Dancelor_client_html
 
-(** Row for when the search returned no results. *)
-let no_results_row =
+(** Generic row showing an emoji on the left and a message on the right. *)
+let emoji_row emoji message =
   tr [
-    td [txt "⚠️"];
-    td [txt "Your search returned no results."];
+    td [txt emoji];
+    td ~a:[a_colspan 4] [txt message];
   ]
 
-(** Rows for when the search returned error messages. *)
-let error_rows messages =
-  Fun.flip List.map messages @@ fun message ->
-  tr [
-    td [txt "❌"];
-    td [txt message];
-  ]
+(** Abstraction of the possible states of the search bar. *)
+type 'a search_bar_state =
+  | StartTyping (** when the user has not typed anything yet *)
+  | ContinueTyping (** when the user has not typed enough yet *)
+  | NoResults (** when the search returned no results *)
+  | Results of 'a list (** when the search returned results; guaranteed to be non empty; otherwise [NoResults] *)
+  | Errors of string list (** when the search returned an error; guaranteed to be non empty *)
 
-let make ~placeholder ~search ~make_result ~max_results ~on_enter =
-  let (search_text, set_search_text) = S.create "" in
+let make ~placeholder ~search ~make_result ~max_results ?on_enter ?(autofocus=false) () =
+  let (search_text, set_search_text_immediately) = S.create "" in
   let (table_visible, set_table_visible) = S.create false in
 
+  (* REVIEW: maybe this should become a generic signal helper to delay signals
+     by a certain time? *)
+  (** The following is a mechanism to only set the search text in a delayed
+      fashion. The [search_text_setter] signal contains an Lwt promise whose job
+      is to set the search text after a delayed time. The [set_search_text]
+      function cancels the current Lwt promise (which does nothing if it already
+      resolved) and creates a new one in its place. *)
+  let (search_text_setter, set_search_text_setter) = S.create Lwt.return_unit in
+  let set_search_text text =
+    (* try cancelling the current search text setter *)
+    Lwt.cancel (S.value search_text_setter);
+    (* prepare the new search text setter *)
+    let new_search_text_setter =
+      (* FIXME: here, we need to delay by something like 100ms but [Lwt_unix]
+         does not seem to be the answer. *)
+      Lwt.pmsleep 0.30;%lwt
+      set_search_text_immediately text;
+      Lwt.return_unit
+    in
+    (* register it in the signal *)
+    set_search_text_setter new_search_text_setter;
+    (* fire it asynchronously *)
+    Lwt.async (fun () -> new_search_text_setter)
+  in
+
+  (** Minimum number of characters for the search to fire. *)
+  let min_characters = 3 in
+  let min_characters_text = "three" in
```

Comment:
Nit: To avoid inconsistencies in the (unlikely) case where min_characters would change in the future, maybe use to_english_string?